blockchain/src/codec.rs

```diff
@@ -0,0 +1,394 @@
+                let shortid_nonce = read(src.read_u64(), "shortid_nonce")?;
+                Ok(Message::GetInventory(GetInventory {
+                    version,
+                    shortid_nonce,
+                }))
+            }
+            MessageType::MempoolTxs => {
+                let tip = read(src.read_blockid(), "inventory version")?;
+                let txs = read(src.read_txs(), "txs")?;
+                Ok(Message::MempoolTxs(MempoolTxs { tip, txs }))
+            }
+            MessageType::GetMempoolTxs => {
+                let shortid_nonce = read(src.read_u64(), "shortid_nonce")?;
+                let shortid_list = read(src.read_shortid_vec(), "shortid_list")?;
+                Ok(Message::GetMempoolTxs(GetMempoolTxs {
+                    shortid_nonce,
+                    shortid_list,
+                }))
+            }
+        }
+    }
+}
+
+impl Encodable for Message {
+    type Error = Error;
+
+    fn encode(&self, dst: &mut impl Writer) -> Result<(), Self::Error> {
+        encode_message(self, dst).map_err(|e| {
+            Error(match e {
+                WriteError::InsufficientCapacity => String::from("insufficient capacity"),
+            })
+        })
+    }
+
+    fn encoded_length(&self) -> usize {
+        unimplemented!()
+    }
+}
+
+fn encode_message(msg: &Message, dst: &mut impl Writer) -> Result<(), WriteError> {
+    match msg {
+        Message::Block(b) => {
+            dst.write_u8(b"message type", MessageType::Block as u8)?;
+            dst.write_block_header(&b.header)?;
+            dst.write_signature(&b.signature)?;
+            dst.write_txs(&b.txs)?;
+        }
+        Message::GetBlock(g) => {
+            dst.write_u8(b"message type", MessageType::GetBlock as u8)?;
+            dst.write_u64(b"block height", g.height)?;
+        }
+        Message::Inventory(inv) => {
+            dst.write_u8(b"message type", MessageType::Inventory as u8)?;
+            dst.write_inventory(&inv)?;
+        }
+        Message::GetInventory(g) => {
+            dst.write_u8(b"message type", MessageType::GetInventory as u8)?;
+            dst.write_u64(b"version", g.version)?;
+            dst.write_u64(b"shortid_nonce", g.shortid_nonce)?;
+        }
+        Message::MempoolTxs(mempool) => {
+            dst.write_u8(b"message type", MessageType::MempoolTxs as u8)?;
+            dst.write_blockid(b"tip", &mempool.tip)?;
+            dst.write_txs(&mempool.txs)?;
+        }
+        Message::GetMempoolTxs(g) => {
+            dst.write_u8(b"message type", MessageType::GetMempoolTxs as u8)?;
+            dst.write_u64(b"shortid_nonce", g.shortid_nonce)?;
+            dst.write_shortid_vec(b"shortid_list", &g.shortid_list)?;
+        }
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{utreexo, BlockHeader, BlockID, BlockTx};
+    use curve25519_dalek::ristretto::CompressedRistretto;
+    use curve25519_dalek::scalar::Scalar;
+    use p2p::reexport::BytesMut;
```

Comment:
I think we can use `Vec<u8>` to test encoding, no need to pull big dependency (even if you'd make it a dev-dependency).

Reply:
p2p was removed from dependencies in blockchain